docs/user-guide/rug/tests.md

```diff
@@ -201,48 +358,220 @@ Given("a Java source file", (p: Project) => {
+look like this:
+
+```typescript
+import { ScenarioWorld } from "@atomist/rug/test/ScenarioWorld";
+import { When } from "@atomist/rug/test/handler/Core";
+
+When("the scenario aborts", (w: ScenarioWorld) => {
+    w.abort();
+});
+```
+
+If you are testing editors and generators, it would look like this:
+
+```typescript
+import { Project } from "@atomist/rug/model/Project";
+import { ScenarioWorld } from "@atomist/rug/test/ScenarioWorld";
+import { When } from "@atomist/rug/test/project/Core";
+
+When("the scenario aborts", (p: Project, w: ScenarioWorld) => {
+    w.abort();
+});
+```
 
 ## Assertions
 
 The `#!gherkin Then` steps consist of one or more assertions about the
-final state of the project.  It is good practice for each step to be
-fine-grained, containing only a single assertion, so that reports
-are maximally informative about what succeeded and failed.  The code
-of each failed assertion will be available in the test report provided
-by the Rug CLI.
-
-The callback provided as the second argument to the `#!typescript
-Then` accepts the `#!typescript Project` and optional `#!typescript
-ScenarioWorld` as arguments and returns a `#!typescript boolean`.
-Often the assertions use the `#!typescript Project.fileExists()` and
-`#!typescript Project.fileContains()` methods.  For example, a typical
+final state of the scenario world.  The `#!typescript Then`
+implementation should return either `#!typescript boolean` or
+`#!typescript void`.  It is good practice for each step to be
+fine-grained, containing only a single assertion, so that reports are
+maximally informative about what succeeded and failed.  The code of
+each failed assertion will be available in the test report provided by
+the Rug CLI.
+
+The `#!typescript ScenarioWorld` interface provides methods for seeing
+if parameters passed to a Rug were valid and checking if a scenario
+aborted, e.g., if a Rug throws an exception.  The example below shows
+how to use each of these methods in a handler scenario `#!gherkin
+Then` step.
+
+```typescript
+import { ScenarioWorld } from "@atomist/rug/test/ScenarioWorld";
+import { Then } from "@atomist/rug/test/handler/Core";
+
+Then("the parameters were valid", (w: ScenarioWorld) => {
+    return w.invalidParameters() == null;
+});
+
+Then("it aborted", (w: ScenarioWorld) => {
+    return w.aborted();
+});
+```
+
+The code would look similar for editor and generator scenario
+`#!gherkin Then` steps, except that you would import `#!typescript
+Then` from `@atomist/rug/test/project/Core` and each callback would
+accept the arguments `#!typescript (p: Project, w: ScenarioWorld)`.
+But it is probably just easier to use the well-known steps when
+testing editors and generators.
+
+```gherkin
+Then parameters were valid
+Then parameters were invalid
+Then the scenario aborted
+```
+
+### Handler assertions
+
+Handlers return a `#!typescript Plan` object, which is simply a data
```

Comment:
Is it worth calling out that Event Handlers return EventPlans and Command Handlers return CommandPlans?